packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+    const session = this.getSession(origin);
+    assert(session, new Error('No session to close'));
+
+    await this.notify(origin, { method: 'multichainHack_metamask_disconnect' });
+
+    this.update((state) => {
+      delete state.sessions[origin];
+    });
+
+    await Promise.all(
+      Object.values(session.handlingSnaps).map((snapId) =>
+        this.messagingSystem.call('SnapController:onSessionClose', snapId),
+      ),
+    );
+  }
+
+  async onConnect(
+    origin: string,
+    connection: ConnectArguments,
+  ): Promise<Session> {
+    const existingSession = this.getSession(origin);
+    if (existingSession) {
+      await this.closeSession(origin);
+    }
+
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = snaps.filter((snap) => snap.enabled && !snap.blocked);
+
+    const availableNamespaces = await filteredSnaps.reduce<
+      Promise<Record<SnapId, Record<NamespaceId, Namespace>>>
+    >(async (previousPromise, snap) => {
+      const acc = await previousPromise;
+      const snapNamespaces = await this.snapToNamespaces(snap);
+      if (snapNamespaces !== null) {
+        return { ...acc, [snap.id]: snapNamespaces };
+      }
+      return acc;
+    }, Promise.resolve({}));
+
+    // The magical matching algorithm specified in SIP-2
+    const namespaceToSnaps = findMatchingKeyringSnaps(
+      connection.requiredNamespaces,
+      availableNamespaces,
+    );
+
+    const permissions = await this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    );
+
+    const approvedNamespacesAndSnaps = Object.entries(namespaceToSnaps).reduce(
+      (acc, [namespace, snapIds]) => {
+        // If snap already is approved for use, solve conflict by using that snap
+        const approvedSnap = snapIds.find((snapId) => {
+          return (
+            permissions &&
+            hasProperty(permissions, getSnapPermissionName(snapId))
+          );
+        });
+
+        if (approvedSnap) {
+          return { ...acc, [namespace]: [approvedSnap] };
+        }
+        return acc;
+      },
+      {},
+    );
+
+    const hasConflicts = Object.keys(namespaceToSnaps).some(
+      (namespace) => !hasProperty(approvedNamespacesAndSnaps, namespace),
+    );
+
+    const filteredNamespacesAndSnaps = hasConflicts
+      ? namespaceToSnaps
+      : approvedNamespacesAndSnaps;
+
+    const possibleAccounts = await this.namespacesToAccounts(
+      origin,
+      filteredNamespacesAndSnaps,
+      connection.requiredNamespaces,
+    );
+
+    // If current configuration doesn't solve request, we need to show a prompt.
+    const resolvedAccounts = hasConflicts
+      ? await this.resolveConflicts(origin, possibleAccounts)
+      : possibleAccounts;
+    const chosenAccounts = fromEntries(
+      Object.entries(resolvedAccounts).map(([namespace, snapIds]) => [
+        namespace,
+        snapIds[0] ?? null,
```

Comment:
I assume `snapIds` is always `undefined | [AccountId[]]` at this point. Otherwise `this.resolveConflicts` should always just leave it with a singular snap

Reply:
This code is moved from `resolveConflicts`. `resolveConflicts` is now only used when prompting is needed.

Reply:
Oh, you are right! I had forgot how I typed it.